src/reflect/scala/reflect/internal/tpe/TypeMaps.scala

```diff
@@ -834,9 +834,28 @@ private[internal] trait TypeMaps {
   /** Note: This map is needed even for non-dependent method types, despite what the name might imply.
     */
   class InstantiateDependentMap(params: List[Symbol], actuals0: List[Type]) extends TypeMap with KeepOnlyTypeConstraints {
-    private val actuals      = actuals0.toIndexedSeq
-    private val existentials = new Array[Symbol](actuals.size)
-    def existentialsNeeded: List[Symbol] = existentials.iterator.filter(_ ne null).toList
+    private[this] var _actuals: Array[Type] = _
+    private[this] var _existentials: Array[Symbol] = _
+    private def actuals: Array[Type] = {
+      if (_actuals eq null) {
+        val temp = new Array[Type](actuals0.size)
```

Comment:
1. This traverses the list already - is the goal to avoid going through ArrayBuffer?
2. Why not directly assign to `_actuals`?

Reply:
Yeah, the goal was to avoid any intermediate objects. So you either need to guess and re-double (like `ArrayBuffer`), or pre-traverse the list and count the exact size. I opted for the latter.

Reply:
I think wrote it this way working towards to elimate bounds checks on the array writes, but got it wrong. In addition to working on a local variable, I should also terminate the loop on `i < temp.length`. Then JIT can prove all array accesses are in bounds.

Reply:
(This is all a bit of a micro-optimization after the other change, which means we get here less frequently.)

Reply:
pushed the modified version